### ISSUE TITLE
[0.7] Contextmenu not supported by : org.fxmisc.richtext.CodeArea

### ISSUE THREAD

**Issue by pierrecochet** [TomasMikula/RichTextFX#363]
Hi,
Since i use the 0.7 the property ContextMenu isn't recognized, whereas it works on 0.6. Do i have any way to fix it, or any altrenative ?

**Comment by JordanMartinez** [comment]
Just a guess, but I think this is because `StyledTextArea` does not extend `Control` but `Region` in the new release....

**Comment by rivagti** [comment]
@JordanMartinez yeah, you're right. but any idea how to implement it?

**Comment by nazmuddin** [comment]
@JordanMartinez  , I did it like this in `GenericStyledArea`

`private Popup popup;
    private Button cut;
    private Button copy;
    private Button paste;
    private Button select;

    private EventHandler contextMenu = new EventHandler<MouseEvent>() {
        @Override
        public void handle(MouseEvent event) {
            if (event.getButton() == event.getButton().SECONDARY) {
                popup.setAutoFix(false);
                popup.show(area, event.getScreenX(), event.getScreenY());
            } else if (popup.isShowing() && event.getClickCount() == 1) {
                popup.hide();
            }
        }
    };
    
    public RichTextFx() {
        StageFactory fact = StageFactory.INSTANCE;
        mainStage = fact.createStage();

        popup = new Popup();
        cut = new Button("Cut");
        cut.setOnAction(e ->{
            area.cut();
            popup.hide();
        });
        copy = new Button("Copy");
        copy.setOnAction(e -> {
            area.copy();
            popup.hide();
        });
        paste = new Button("Paste");
        paste.setOnAction(e ->{
            area.paste();
            popup.hide();
        });
        select = new Button("Select All");
        select.setOnAction(e->{
            area.selectAll();
        });
        VBox box = new VBox();
        box.setId("popup"); 
        box.getStylesheets().add(getClass().getResource("context-menu.css").toExternalForm());
        box.setPrefSize(80, 50);
        box.getChildren().addAll(cut, copy, paste, select);

        popup.getContent().add(box);
        area.setOnMouseClicked(contextMenu);
   }
`
area is instance of GenericStyledArea. Hope it helps.

**Comment by JordanMartinez** [comment]
@rivagti See @nazmuddin's example in combination with mine below. He uses a `Popup` whereas I use a `ContextMenu`. He uses `setOn[EventType]` whereas I use `InputMap`. The two should give you a good enough idea for how to resolve this issue:
````java
package org.fxmisc.richtext.demo;

import javafx.application.Application;
import javafx.scene.Scene;
import javafx.scene.control.ContextMenu;
import javafx.scene.control.MenuItem;
import javafx.scene.input.MouseButton;
import javafx.stage.Stage;
import org.fxmisc.richtext.InlineCssTextArea;
import org.fxmisc.wellbehaved.event.EventPattern;
import org.fxmisc.wellbehaved.event.InputMap;
import org.fxmisc.wellbehaved.event.Nodes;

public class ContextMenuExample extends Application {

    @Override
    public void start(Stage primaryStage) {
        // create your context menu somewhere in your code...
        ContextMenu menu = new ContextMenu();
        menu.getItems().addAll(
                new MenuItem("Menu 1"),
                new MenuItem("Menu 2"),
                new MenuItem("Menu 3")
        );

        InlineCssTextArea area = new InlineCssTextArea("Some text");

        // add an input mapping that shows that context menu when you 
        //  right click somewhere in the area
        Nodes.addInputMap(area, 
                InputMap.consume(
                    EventPattern.mouseClicked(MouseButton.SECONDARY),
                    e -> {
                        // show the area using the mouse event's screen x & y values
                        menu.show(area, e.getScreenX(), e.getScreenY());
                    }
                )
        );

        primaryStage.setScene(new Scene(area, 500, 500));
        primaryStage.show();
    }

    public static void main(String[] args) {
        launch(args);
    }
}
````

**Comment by JordanMartinez** [comment]
Should this feature be supported out-of-box? 
We could add these methods in `StyledTextArea`:
````java
private ContextMenu contextMenu = null
public final Optional<ContextMenu> getContextMenu() {
    return Optional.ofNullable(contextMenu);
}
public final void setContextMenu(ContextMenu menu) {
    contextMenu = menu;
}

private double contextMenuOffsetX;
public final double getContextMenuOffsetX() { return contextMenuOffsetX; }
public final void setContextMenuOffsetX(double xOffset) { contextMenuOffsetX = xOffset; }

// same for yOffset
````
and add this handler to `StyledTextAreaBehavior`
````java
InputMapTemplate.consumeUnless(
    mouseClicked(SECONDARY),
    () -> !area.getContextMenu().isPresent(),
    e -> {
            double xOffset = area.getContextMenuOffsetX();
            double yOffset = area.getContextMenuOffsetY();
            area.getContextMenu().show(area, e.getScreenX() + xOffset, e.getScreenY() + yOffset);
        }
}
````

**Comment by rivagti** [comment]
@JordanMartinez yes, I think so. anyway thank you both... @nazmuddin

**Comment by berry120** [comment]
I ended up here after searching for how to get this working in 0.7, so +1 for supporting this out the box (especially since it's a feature regression from 0.6 otherwise.)

**Comment by JordanMartinez** [comment]
Closed by #439.
